Description: add username to lock dialog and adjust fonts and spacing
 when used with Unity
Author: Marc Deslauriers <marc.deslauriers@canonical.com>
Forwarded: no, Ubuntu-specific

Index: gnome-screensaver/src/gs-lock-plug.c
===================================================================
--- gnome-screensaver.orig/src/gs-lock-plug.c	2012-02-25 21:33:54.000000000 +0100
+++ gnome-screensaver/src/gs-lock-plug.c	2012-03-07 15:13:56.989846836 +0100
@@ -72,6 +72,7 @@
 
         GtkWidget   *notebook;
         GtkWidget   *auth_face_image;
+        GtkWidget   *auth_realname_label;
         GtkWidget   *auth_prompt_label;
         GtkWidget   *auth_prompt_entry;
         GtkWidget   *auth_prompt_box;
@@ -1254,9 +1255,18 @@
         gtk_widget_set_sensitive (plug->priv->auth_unlock_button, TRUE);
         gtk_widget_show (plug->priv->auth_unlock_button);
         gtk_widget_grab_default (plug->priv->auth_unlock_button);
-        markup = g_strdup_printf ("<b><big>%s</big></b>", message);
+
+        /* Change appearance if we're running under Unity */
+        if (g_getenv ("XDG_CURRENT_DESKTOP") &&
+            strcmp (g_getenv ("XDG_CURRENT_DESKTOP"), "Unity") == 0) {
+                markup = g_strdup_printf ("<span font_desc=\"Ubuntu 10\">%s</span>", message);
+        } else {
+                markup = g_strdup_printf ("<b><big>%s</big></b>", message);
+        }
+
         gtk_label_set_markup (GTK_LABEL (plug->priv->auth_prompt_label), markup);
         g_free (markup);
+
         gtk_widget_show (plug->priv->auth_prompt_label);
         gtk_entry_set_visibility (GTK_ENTRY (plug->priv->auth_prompt_entry), visible);
         gtk_widget_set_sensitive (plug->priv->auth_prompt_entry, TRUE);
@@ -1387,6 +1397,39 @@
         gs_profile_end ("page one buttons");
 }
 
+static char *
+get_user_display_name (void)
+{
+        const char *name;
+        char       *utf8_name;
+
+        name = g_get_real_name ();
+
+        if (name == NULL || strcmp (name, "Unknown") == 0) {
+                name = g_get_user_name ();
+        }
+
+        utf8_name = NULL;
+
+        if (name != NULL) {
+                utf8_name = g_locale_to_utf8 (name, -1, NULL, NULL, NULL);
+        }
+
+        return utf8_name;
+}
+
+static void
+update_realname_label (GSLockPlug *plug)
+{
+        char *name;
+        char *markup;
+        name = get_user_display_name ();
+        markup = g_strdup_printf ("<span font_desc=\"Ubuntu 16\">%s</span>", name);
+        gtk_label_set_markup (GTK_LABEL (plug->priv->auth_realname_label), markup);
+        g_free (markup);
+        g_free (name);
+}
+
 static void
 create_page_one (GSLockPlug *plug)
 {
@@ -1418,25 +1461,59 @@
         gtk_box_pack_start (GTK_BOX (hbox), plug->priv->auth_prompt_kbd_layout_indicator, FALSE, FALSE, 0);
 #endif
 
-        plug->priv->auth_prompt_label = gtk_label_new_with_mnemonic (_("_Password:"));
-        gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_prompt_label), 0, 0.5);
-        gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_prompt_label, FALSE, FALSE, 0);
+        /* Change appearance if we're running under Unity */
+        if (g_getenv ("XDG_CURRENT_DESKTOP") &&
+            strcmp (g_getenv ("XDG_CURRENT_DESKTOP"), "Unity") == 0) {
+                gtk_box_set_spacing (vbox2, 0);
+                gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_face_image), 0, 0.5);
+                plug->priv->auth_realname_label = gtk_label_new (NULL);
+                update_realname_label (plug);
+                gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_realname_label), 0, 1);
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_realname_label, FALSE, FALSE, 0);
+
+                plug->priv->auth_prompt_label = gtk_label_new_with_mnemonic (_("_Password:"));
+                gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_prompt_label), 0, 0);
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_prompt_label, FALSE, FALSE, 0);
+
+                plug->priv->auth_prompt_entry = gtk_entry_new ();
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_prompt_entry, TRUE, TRUE, 6);
+
+                gtk_label_set_mnemonic_widget (GTK_LABEL (plug->priv->auth_prompt_label),
+                                               plug->priv->auth_prompt_entry);
+
+                plug->priv->auth_capslock_label = gtk_label_new ("");
+                gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_capslock_label), 0, 0.5);
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_capslock_label, FALSE, FALSE, 0);
+
+                /* Status text */
+
+                plug->priv->auth_message_label = gtk_label_new (NULL);
+                gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_message_label), 0, 0.5);
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_message_label,
+                                    FALSE, FALSE, 0);
+
+        } else {
+
+                plug->priv->auth_prompt_label = gtk_label_new_with_mnemonic (_("_Password:"));
+                gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_prompt_label), 0, 0.5);
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_prompt_label, FALSE, FALSE, 0);
 
-        plug->priv->auth_prompt_entry = gtk_entry_new ();
-        gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_prompt_entry, TRUE, TRUE, 0);
+                plug->priv->auth_prompt_entry = gtk_entry_new ();
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_prompt_entry, TRUE, TRUE, 0);
 
-        gtk_label_set_mnemonic_widget (GTK_LABEL (plug->priv->auth_prompt_label),
-                                       plug->priv->auth_prompt_entry);
+                gtk_label_set_mnemonic_widget (GTK_LABEL (plug->priv->auth_prompt_label),
+                                           plug->priv->auth_prompt_entry);
 
-        plug->priv->auth_capslock_label = gtk_label_new ("");
-        gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_capslock_label), 0.5, 0.5);
-        gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_capslock_label, FALSE, FALSE, 0);
+                plug->priv->auth_capslock_label = gtk_label_new ("");
+                gtk_misc_set_alignment (GTK_MISC (plug->priv->auth_capslock_label), 0.5, 0.5);
+                gtk_box_pack_start (GTK_BOX (vbox2), plug->priv->auth_capslock_label, FALSE, FALSE, 0);
 
-        /* Status text */
+                /* Status text */
 
-        plug->priv->auth_message_label = gtk_label_new (NULL);
-        gtk_box_pack_start (GTK_BOX (vbox), plug->priv->auth_message_label,
-                            FALSE, FALSE, 0);
+                plug->priv->auth_message_label = gtk_label_new (NULL);
+                gtk_box_pack_start (GTK_BOX (vbox), plug->priv->auth_message_label,
+                                    FALSE, FALSE, 0);
+        }
         /* Buttons */
         plug->priv->auth_action_area = gtk_button_box_new (GTK_ORIENTATION_HORIZONTAL);
 
